include/dd/NoiseFunctionality.hpp

```diff
@@ -0,0 +1,714 @@
+    class StochasticNoiseFunctionality {
+    public:
+        StochasticNoiseFunctionality(const std::unique_ptr<DDPackage>& package,
+                                     dd::QubitCount                    nQubits,
+                                     double                            gateNoiseProbability,
+                                     double                            amplitudeDampingProb,
+                                     double                            multiQubitGateFactor,
+                                     std::vector<dd::NoiseOperations>  noiseEffects):
+            package(package),
+            nQubits(nQubits),
+            dist(0.0, 1.0L),
+            noiseProbability(gateNoiseProbability),
+            noiseProbabilityMulti(gateNoiseProbability * multiQubitGateFactor),
+            sqrtAmplitudeDampingProbability({std::sqrt(amplitudeDampingProb), 0}),
+            oneMinusSqrtAmplitudeDampingProbability({std::sqrt(1 - amplitudeDampingProb), 0}),
+            sqrtAmplitudeDampingProbabilityMulti({std::sqrt(gateNoiseProbability) * multiQubitGateFactor, 0}),
+            oneMinusSqrtAmplitudeDampingProbabilityMulti({std::sqrt(1 - multiQubitGateFactor * amplitudeDampingProb), 0}),
+            noiseEffects(std::move(noiseEffects)) {
+            ampDampingFalse      = dd::GateMatrix({dd::complex_one, dd::complex_zero, dd::complex_zero, oneMinusSqrtAmplitudeDampingProbability});
+            ampDampingFalseMulti = dd::GateMatrix({dd::complex_one, dd::complex_zero, dd::complex_zero, oneMinusSqrtAmplitudeDampingProbabilityMulti});
+            ampDampingTrue       = dd::GateMatrix({dd::complex_zero, sqrtAmplitudeDampingProbability, dd::complex_zero, dd::complex_zero});
+            ampDampingTrueMulti  = dd::GateMatrix({dd::complex_zero, sqrtAmplitudeDampingProbabilityMulti, dd::complex_zero, dd::complex_zero});
+            identityDD           = package->makeIdent(nQubits);
+            package->incRef(identityDD);
+        }
+
+        ~StochasticNoiseFunctionality() {
+            package->decRef(identityDD);
+        }
+
+    protected:
+        const std::unique_ptr<DDPackage>&      package;
+        dd::QubitCount                         nQubits;
+        std::uniform_real_distribution<dd::fp> dist;
+
+        double                           noiseProbability;
+        double                           noiseProbabilityMulti;
+        dd::ComplexValue                 sqrtAmplitudeDampingProbability;
+        dd::ComplexValue                 oneMinusSqrtAmplitudeDampingProbability;
+        dd::ComplexValue                 sqrtAmplitudeDampingProbabilityMulti;
+        dd::ComplexValue                 oneMinusSqrtAmplitudeDampingProbabilityMulti;
+        dd::GateMatrix                   ampDampingTrue{};
+        dd::GateMatrix                   ampDampingTrueMulti{};
+        dd::GateMatrix                   ampDampingFalse{};
+        dd::GateMatrix                   ampDampingFalseMulti{};
```

Comment:
technically, these should be const as well, but then you have to change the constructor slightly.